src/core/thread/network_data_notifier.cpp

```diff
@@ -152,6 +169,75 @@ void Notifier::HandleCoapResponse(Error aResult)
     }
 }
 
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE && OPENTHREAD_CONFIG_BORDER_ROUTER_REQUEST_ROUTER_ROLE
+
+bool Notifier::IsEligibleForRouterRoleUpgrade(void) const
+{
+    bool    isEligible = false;
+    uint8_t activeRouterCount;
+
+    VerifyOrExit(Get<Mle::MleRouter>().IsChild() && Get<Mle::MleRouter>().IsRouterEligible());
+
+    // RouterUpgradeThreshold can be explicitly set to zero in some of
+    // cert tests to disallow device to become router.
+
+    VerifyOrExit(Get<Mle::MleRouter>().GetRouterUpgradeThreshold() != 0);
+
+    // Check the local network data to ensure that we are a border
+    // router providing IP connectivity and therefore eligible to
+    // request router role upgrade with `kBorderRouterRequst`
+    // status.
+
+    VerifyOrExit(Get<Local>().CountBorderRouters(kAnyRole) > 0);
```

Comment:
why don't we count the number of BR entries advertised by this device in Leader's Network Data?

Reply:
This does the same (`NetworkData::Local` conatains all the stuff this device wants to register in network data)
Also we want to try to become router before we may register the entries (to avoid re-registering again once becoming leader).

Reply:
I doubt if we can become Router before we may register the entries because the entries will be synchronized to Leader  very soon, meanwhile the new BR Router request will be sent after a few seconds. 

So, I think there is no notable difference if the local data can be synced to Leader.
Is it better to treat Leader Network Data as the single source of truth, which should also simplify calls to ScheduleRouterRoleUpgradeIfEligible.

Reply:
I agree. I think @EskoDijk has a very good point on this (related to leader behavior in accepting/rejecting such request). Changed this in the new push (we check `Leader` network data to see if the device's rloc16 is present in it and also it meets the requirement for being counted as br providing IP connectivity).

Reply:
Thanks @EskoDijk for the lucid elaboration. 
I think it's a general convention to use the Leader's Network Data as the SSOT, and use the Local Network Data as the intent, to avoid tricky issues.

Reply:
Just a thought: to be precise, the BR would have to check that the Leader's network data indicates that the BR is "eligible for upgrade" (i.e. providing connectivity) and also its own local version of network data says that the BR is "eligible for upgrade".  Only then does it make sense to request for upgrade.    The Leader's version of network data may be outdated, or not matching the intent  ;-)

Reply:
@EskoDijk interesting and good point. It is a corner case, but I think it is easy to handle and support. I updated `IsEligibleForRouterRoleUpgradeAsBorderRouter()` to now check in both local network data and in leader network data.

Reply:
I think `IsEligibleForRouterRoleUpgradeAsBorderRouter` has handled these cases correctly.